### PR TITLE
feat: spectrogram loading optimization with state machine

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

- **New spectrogram loader state machine** (`spectrogramLoader.svelte.ts`) that decouples status checking from image fetching, preventing head-of-line blocking in the browser's HTTP/1.1 connection pool
- **Distinct UI feedback** showing "Waiting..." when queued on the server and "Generating..." when actively being processed, instead of a generic spinner
- **Permanent failure protection** preventing infinite retry loops when the server reports a spectrogram as permanently failed
- **Adaptive backend concurrency** using `runtime.NumCPU()` instead of a hardcoded limit of 4 for spectrogram generation

## Problem

On slower hardware (Raspberry Pi), spectrogram generation via SoX can take several seconds. The previous implementation used image queue slots for both status checking and image loading, causing head-of-line blocking — slow generations would occupy all available slots, preventing already-cached spectrograms from loading promptly.

## Approach

The new state machine follows a status-check-first pattern:
1. **Check status** via lightweight JSON endpoint (no queue slot needed)
2. **If cached**: acquire queue slot and load image immediately
3. **If not started**: trigger generation, then poll with exponential backoff
4. **If generating**: poll with exponential backoff until ready
5. **Only acquire a queue slot** when the image is confirmed ready

This keeps the HTTP/1.1 connection budget (6 per domain, minus 1 for SSE = 5) optimally allocated: 3 slots for image loading + 2 for status checks and API calls.

## Backend change

`maxConcurrentSpectrograms` in `media.go` changed from `const 4` to `var runtime.NumCPU()`, so the spectrogram generation semaphore adapts to actual hardware capabilities.

## Test plan

- [x] 25 new unit tests covering all state transitions, error handling, cleanup, and edge cases
- [x] All 1656 frontend tests pass
- [x] `npm run check:all` clean
- [x] `go build ./...` succeeds
- [x] `golangci-lint run -v` passes with 0 issues

## Summary by CodeRabbit

* **New Features**
  * Visibility-triggered spectrogram loading with a centralized loader and updated Waiting/Generating indicators.

* **Performance**
  * Spectrogram generation concurrency now scales with available CPU cores.

* **Bug Fixes / Reliability**
  * Improved retry, polling, error handling, and cleanup to reduce stuck, leaking, or failed loads.

* **Tests**
  * Added comprehensive tests covering loader flows, retries, polling, lifecycle, and error scenarios.

### CODE REVIEW

File: frontend/src/lib/desktop/features/dashboard/components/DetectionCard.svelte [hunk 1]
```diff
@@ -304,24 +154,32 @@
     document.body.removeChild(link);
   }
 
+  // eslint-disable-next-line no-undef -- browser global
+  let observer: IntersectionObserver | undefined;
```
[the remark this answers is not in the document]
Comment (tphakala) [reply frontend/src/lib/desktop/features/dashboard/components/DetectionCard.svelte:158]: Not duplicates. Line 157 suppresses `no-undef` for the `IntersectionObserver` **type annotation**, and line 163 suppresses it for the `new IntersectionObserver` **constructor call**. Each `eslint-disable-next-line` only covers the immediately following line, so both are needed.

File: frontend/src/lib/utils/spectrogramLoader.svelte.ts [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+  async function acquireAndLoad(detectionId: number): Promise<void> {
+    if (isStale(detectionId)) return;
+
+    state = 'acquiring-slot';
+
+    slotHandle = acquireSlot();
+    const acquired = await slotHandle.promise;
+
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- state can change during await
+    if (!acquired || isStale(detectionId) || state !== 'acquiring-slot') {
```
[the remark this answers is not in the document]
Comment (tphakala) [reply frontend/src/lib/utils/spectrogramLoader.svelte.ts:356]: The `eslint-disable-next-line` directive is required here — without it, TypeScript's control flow analysis considers the `state \!== 'acquiring-slot'` check redundant (since it was just set to `'acquiring-slot'` above). The `await` on line 340 yields, allowing other code to change `state`, but TypeScript doesn't model this. Removing the directive would cause a lint error.

File: frontend/src/lib/utils/spectrogramLoader.svelte.ts [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+        config.imageRetryDelays[Math.min(imageRetryCount, config.imageRetryDelays.length - 1)];
+      imageRetryCount++;
+
+      logger.debug('Spectrogram image load failed, retrying', {
+        detectionId,
+        retryCount: imageRetryCount,
+        delay,
+      });
+
+      pollTimer = setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (tphakala) [reply frontend/src/lib/utils/spectrogramLoader.svelte.ts:433]: Fixed in 971644d. Slot is now released immediately on image error. Retries go through `acquireAndLoad` to re-acquire a slot, so slots aren't held idle during retry delays.

File: frontend/src/lib/utils/spectrogramLoader.svelte.ts [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+  function handleImageError(): void {
+    const detectionId = currentDetectionId;
+    if (detectionId === undefined) return;
+
+    // Release slot immediately so other cards can load during retry delay
+    releaseCurrentSlot();
+
+    if (imageRetryCount < config.maxImageRetries) {
+      const delay =
+        config.imageRetryDelays[Math.min(imageRetryCount, config.imageRetryDelays.length - 1)];
```
[the remark this answers is not in the document]
Comment (tphakala) [reply frontend/src/lib/utils/spectrogramLoader.svelte.ts:423]: Fixed in 83852b8. Added timestamp cache-buster to `buildImageUrl` when `imageRetryCount > 0`.

File: frontend/src/lib/utils/spectrogramLoader.svelte.ts [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+    return `/api/v2/spectrogram/${detectionId}/status?size=${config.size}&raw=${String(config.raw)}`;
+  }
+
+  function buildGenerateUrl(detectionId: number): string {
+    return `/api/v2/spectrogram/${detectionId}/generate?size=${config.size}&raw=${String(config.raw)}`;
+  }
+
+  function buildImageUrl(detectionId: number): string {
+    return `/api/v2/spectrogram/${detectionId}?size=${config.size}&raw=${String(config.raw)}`;
+  }
```
[the remark this answers is not in the document]
Comment (tphakala) [reply frontend/src/lib/utils/spectrogramLoader.svelte.ts:159]: Fixed in 83852b8. Used the same approach — `buildImageUrl` now appends `&t=${Date.now()}` when `imageRetryCount > 0`.

File: frontend/src/lib/utils/spectrogramLoader.svelte.ts [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+        if (fetchRetries > config.maxFetchRetries) {
+          logger.error('Status check failed after retries', { detectionId, error: err });
+          state = 'error';
+          hasError = true;
+          showSpinner = false;
+          return;
+        }
+        await new Promise(resolve => setTimeout(resolve, 500 * fetchRetries));
+      }
+    }
```
[the remark this answers is not in the document]
Comment (tphakala) [reply frontend/src/lib/utils/spectrogramLoader.svelte.ts:232]: Fixed in 76fe3bd. Moved `new AbortController()` inside the retry loop and added `isStale` guard after the delay.

File: frontend/src/lib/utils/spectrogramLoader.svelte.ts [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+    if (isStale(detectionId)) return;
+
+    state = 'triggering';
+    abortController = new AbortController();
+
+    try {
+      const response = await fetch(buildGenerateUrl(detectionId), {
+        method: 'POST',
+        signal: abortController.signal,
+      });
```
[the remark this answers is not in the document]
Comment (tphakala) [reply frontend/src/lib/utils/spectrogramLoader.svelte.ts:252]: Fixed in 76fe3bd. Added `getCsrfToken()` from `$lib/utils/api` and sends `X-CSRF-Token` header on the POST request.

File: frontend/src/lib/utils/spectrogramLoader.svelte.ts [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+        config.imageRetryDelays[Math.min(imageRetryCount, config.imageRetryDelays.length - 1)];
+      imageRetryCount++;
+
+      logger.debug('Spectrogram image load failed, retrying', {
+        detectionId,
+        retryCount: imageRetryCount,
+        delay,
+      });
+
+      // Re-acquire slot through normal flow on retry
```
[the remark this answers is not in the document]
Comment (tphakala) [reply frontend/src/lib/utils/spectrogramLoader.svelte.ts:432]: Already fixed in 83852b8 — `buildImageUrl` appends `&t=${Date.now()}` when `imageRetryCount > 0`.

File: frontend/src/lib/utils/spectrogramLoader.test.ts [hunk 9]
```diff
@@ -0,0 +1,526 @@
...
+      loader.start(42);
+      await flushAll();
+
+      expect(mockFetch).toHaveBeenCalledTimes(1);
+      loader.destroy();
+    });
+  });
+
+  describe('image error handling', () => {
+    it('retries on image error with cache buster', async () => {
```
[the remark this answers is not in the document]
Comment (tphakala) [reply frontend/src/lib/utils/spectrogramLoader.test.ts:327]: Test now correctly asserts cache-busting URL with `toContain` and verifies CSRF header.

File: frontend/src/lib/utils/spectrogramLoader.svelte.ts [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+/**
+ * Creates a spectrogram loader for a single detection card.
+ *
+ * Call `start(detectionId)` when the card becomes visible.
+ * Call `stop()` when hidden or destroyed.
+ * Read `state`, `spectrogramUrl`, `showSpinner`, `error` for UI binding.
+ */
+export function createSpectrogramLoader(userConfig: SpectrogramLoaderConfig = {}) {
+  const config = { ...DEFAULT_CONFIG, ...userConfig };
+
```
[the remark this answers is not in the document]
Comment (tphakala) [reply frontend/src/lib/utils/spectrogramLoader.svelte.ts:73]: The parameter is typed `SpectrogramLoaderConfig = {}` — TypeScript enforces the type at compile time and the default is an empty object. This is an internal utility called only from `DetectionCard.svelte` with a literal object. Adding a runtime type guard for an internal function with no external API surface is unnecessary.

**Comment by tphakala** [comment]
All review items addressed:

**Fixed:**
- Slot released immediately on image error to prevent blocking (Sentry)
- Cache-busting parameter on retry URLs (Sentry, Gemini)
- CSRF token on POST generate request (CodeRabbit)
- AbortController created per retry attempt to prevent race condition (CodeRabbit)
- eslint-disable explanations for consistency (Gemini)
- Test description wording (Gemini)

**Pushed back (with technical reasoning):**
- Duplicate eslint-disable comments — not duplicates, each covers a different line (Gemini)
- Remove eslint-disable for unnecessary-condition — required because TypeScript can't model state changes across await (Gemini)
- Runtime type guard on typed internal function — YAGNI for internal utility with TypeScript enforcement (CodeRabbit)

All CI checks green. Ready to merge.
